File: python/day24.py
from aoc import *


def parse_input(data):
    walls = set()
    blizzards = set()
    for y, line in enumerate(data, -1):
        for x, c in enumerate(line, -1):
            match c:
                case '#': walls.add((x, y))
                case '>': blizzards.add((x, y, 1, 0))
                case '<': blizzards.add((x, y, -1, 0))
                case 'v': blizzards.add((x, y, 0, 1))
                case '^': blizzards.add((x, y, 0, -1))
    height, width = len(data)-2, len(data[0])-2
    walls |= {(0, -2), (width-1, height+1)}
    return blizzards, walls, height, width


def traverse(obstacles, h, w, start, goal, t=0):
    blizzards, walls = obstacles
    positions = {start}
    while positions:
        if goal in positions: return t
        t += 1
        candidates = {nb for pt in positions
                         for nb in neighbours(*pt, 5)}
        obstacles = walls | {((x + dx*t) % w, (y + dy*t) % h)
                             for x, y, dx, dy in blizzards}
        positions = candidates - obstacles


def solve(filename=24):
    *obstacles, h, w = parse_input(read_input(filename))
    start = (0, -1)
    goal = (w-1, h)
    p1 = traverse(obstacles, h, w, start, goal)
    back = traverse(obstacles, h, w, goal, start, p1)
    p2 = traverse(obstacles, h, w, start, goal, back)
    return p1, p2


print(solve())










def test():
    assert solve("24_test") == (18, 54)
    assert solve() == (266, 853)

File: clojure/day24.clj
(ns day24
  (:require aoc
            [clojure.data.int-map :as i]))


(def ^:const N 128)
(def ^:const -N (- N))


(defn pt->num ^long [[^long x ^long y]]
  (+ x (* N y)))

(defn neighbours ^longs [^long pt]
  (for [^long delta [-1 1 -N N]]
    (+ pt delta)))

(defn move-blizzards [blizzards w h t]
  (let [w (long w) , h (long h) , t (long t)]
    (map (fn [[[^long x ^long y] [^long dx ^long dy]]]
           (pt->num [(mod (+ x (* dx t)) w)
                     (mod (+ y (* dy t)) h)]))
         blizzards)))

(defn traverse [[blizzards walls w h] start goal t]
  (reduce
   (fn [queue ^long t]
     (if (queue goal) (reduced (dec t))
         (let [candidates (into queue (mapcat neighbours queue))
               obstacles  (into walls (move-blizzards blizzards w h t))]
           (i/difference candidates obstacles))))
   (i/dense-int-set [start])
   (iterate inc t)))


(defn parse-input [input]
  (let [inp (aoc/read-input input :vector)
        res (for [[y line] (map-indexed vector inp)
                  [x char] (map-indexed vector line)
                  :let [pt [(dec x) (dec y)]
                        pt' (pt->num pt)]]
              (case char
                \# [nil         pt']
                \> [[pt [1  0]] nil]
                \< [[pt [-1 0]] nil]
                \v [[pt [0  1]] nil]
                \^ [[pt [0 -1]] nil]
                [nil nil]))
        blizzards (->> res (map first) (remove empty?))
        walls' (->> res (map second) (remove nil?))
        w (mod  (last walls') N)
        h (quot (last walls') N)
        walls (i/dense-int-set (into walls' [(* 2 -N)
                                             (pt->num [(dec w) (inc h)])]))]
    [blizzards walls w h]))

(defn solve
  ([] (solve 24))
  ([input]
   (let [[_ _ w h :as data] (parse-input input)
         start -N
         goal  (pt->num [(dec w) h])
         traverse_ (partial traverse data)
         p1 (->> 0
                 (traverse_ start goal))
         p2 (->> p1
                 (traverse_ goal start)
                 (traverse_ start goal))]
     [p1 p2])))


(solve)

